src/main/java/com/google/collegeplanner/servlets/SchedulerServlet.java

```diff
@@ -0,0 +1,141 @@
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.http.client.utils.URIBuilder;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+/** Servlet that returns a list of possible schedules containing the given courses.*/
+@WebServlet("/api/scheduler")
+public class SchedulerServlet extends BaseServlet {
+  private ArrayList<String> courseList;
+  private ArrayList<ArrayList<Section>> courses;
+
+  /**
+   * Organizes courses from POST request into a given number of semesters
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    JSONObject body;
+    JSONArray selectedClasses;
+    try {
+      body = getPostRequestBody(request);
+      selectedClasses = (JSONArray) body.get("selectedClasses");
+    } catch (NumberFormatException | ClassCastException | ParseException | NullPointerException e) {
+      respondWithError(
+          "Invalid body for POST request.", HttpServletResponse.SC_BAD_REQUEST, response);
+      return;
+    }
+    
+    try {
+      prepareLists(selectedClasses, response);
+    } catch (ParseException e) {
+      return;
+    }
+
+    JSONObject schedules = getSchedules();
+    response.setContentType("application/json;");
+    response.getWriter().println(new Gson().toJson(schedules));
+  }
+
+  /**
+   * This method loads the courses and courseList ArrayLists with correct
+   * information. courses is loaded with the lists of sections for each
+   * course while courseList is loaded with the courseID's of the selected
+   * courses.
+   * @param classes The JSONArray of courseIds of the selected classes
+   * @param response The HttpServletResponse object
+   */
+  private void prepareLists(JSONArray classes, HttpServletResponse response) 
+  throws IOException, ParseException {
+    String courseId;
+    URI uri;
+    JSONArray jsonArray;
+    ApiUtil apiUtil = new ApiUtil();
+
+    courseList = new ArrayList<String>();
+    courses = new ArrayList<ArrayList<Section>>();
+    for (Object obj : classes) {
+      courseId = (String) obj;
+      courseList.add(courseId);
+      try {
+        uri = new URI("https://api.umd.io/v1/courses/" + courseId + "/sections");
+      } catch (URISyntaxException e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+
+      jsonArray = apiUtil.getJsonArray(uri);
+      if (jsonArray == null) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        throw new ParseException(0);
```

Comment:
throwing ParseException here unfortunately isn't quite right. You're not parsing an input as this point. Honestly, it might make sense to create an InvalidApiResponseError exception for you to use in this case. Alternatively, you can return a boolean in this function (true/false) and just return a generic exception for now if that helps.